src/test/java/htsjdk/samtools/fastq/FastqReaderWriterTest.scala

```diff
@@ -0,0 +1,151 @@
+package htsjdk.samtools.fastq
+
+import java.io.{BufferedReader, File, StringReader}
+import java.nio.file.{Path, Paths}
+
+import htsjdk.UnitSpec
+import htsjdk.samtools.SAMUtils
+import htsjdk.samtools.util.IOUtil
+
+import scala.util.Random
+import scala.collection.JavaConverters.asScalaIterator
+
+class FastqReaderWriterTest extends UnitSpec {
+  private val rng = new Random()
+  private val Bases = Array('A', 'C', 'G', 'T')
+
+  /** Generates a random string of bases of the desired length. */
+  def bases(length: Int): String = {
+    val chs = new Array[Char](length)
+    chs.indices.foreach(i => chs(i) = Bases(rng.nextInt(Bases.length)))
+    new String(chs)
+  }
+
+  /** Generates a FastqRecord with random bases at a given length. */
+  def fq(name: String, length: Int, qual: Int = 30): FastqRecord = {
+    new FastqRecord(name, bases(length), "", SAMUtils.phredToFastq(qual).toString * length)
+  }
+
+  "FastqWriter" should "write four lines per record to file" in {
+    val path = makeTempFile("test.", ".fastq")
+    val out = new FastqWriterFactory().newWriter(path.toFile)
+    val recs = Seq(fq("q1", 50), fq("q2", 48), fq("q3", 55))
+    val Seq(q1, q2, q3) = recs
+
+    recs.foreach(rec => out.write(rec))
+    out.close()
+
+    val lines = IOUtil.slurpLines(path.toFile)
+    lines should have size 12
+
+    lines.get(0) shouldBe "@q1"
+    lines.get(1) shouldBe q1.getReadString
+    lines.get(4) shouldBe "@q2"
+    lines.get(5) shouldBe q2.getReadString
+    lines.get(8) shouldBe "@q3"
+    lines.get(9) shouldBe q3.getReadString
+  }
+
+  it should "write a record with only a single base" in {
+    val path = makeTempFile("test.", ".fastq")
+    val out = new FastqWriterFactory().newWriter(path.toFile)
+    out.write(fq("q1", 1))
+    out.close()
+    IOUtil.slurpLines(path.toFile).get(1) should have length 1
+  }
+
+  it should "write a record with zero-length bases and quals" in {
+    val path = makeTempFile("test.", ".fastq")
+    val out = new FastqWriterFactory().newWriter(path.toFile)
+    out.write(fq("q1", 0))
+    out.close()
+    IOUtil.slurpLines(path.toFile).get(1) should have length 0
+  }
+
+
+  "FastqReader" should "read back a fastq file written by FastqWriter" in {
+    val path = makeTempFile("test.", ".fastq")
+    val out = new FastqWriterFactory().newWriter(path.toFile)
+    val recs = Seq(fq("q1", 50), fq("q2", 100), fq("q3", 150))
+    recs.foreach(rec => out.write(rec))
+    out.close()
+
+    val in = new FastqReader(path.toFile)
+    val recs2 = in.iterator().toList
+    in.close()
+    recs2 should contain theSameElementsInOrderAs recs
+  }
+
+  it should "throw an exception if the input fastq is garbled" in {
+    val fastq =
+      """
+        |@q1
+        |AACCGGTT
+        |+
+        |########
+        |@q2
+        |ACGT
+        |####
+      """.stripMargin.trim
+
+    val in = new FastqReader(null, new BufferedReader(new StringReader(fastq)))
+    an[Exception] shouldBe thrownBy { in.next() }
```

Comment:
I think something very funny is going on with codecov.  Both running in IntelliJ, and running `./gradlew test jacocoTestReport` locally show those lines getting covered.  I'm going to try making the other recommended changes, squashing and then opening up a new PR to see if that will get codecov to recognize the changes.